oximeter/db/src/native/io/table_columns.rs

```diff
@@ -0,0 +1,202 @@
+//! <https://github.com/ClickHouse/ClickHouse/blob/424ecb86e0f04a9422dd01f9216cdbe3c0e314e2/src/Storages/ColumnsDescription.cpp#L132>,
+//!
+//! > NOTE: Serialization format is insane.
+//!
+//! The server serializes an array of `ColumnDescription`s to help the client
+//! validate data before inserting it. This is a pure-text format that includes
+//! column names and types, but also comments, codecs, and TTL values. The
+//! format is indeed pretty wonky, so we're using `nom` to help parse it
+//! robustly. For now, we only care about the names and data types; the
+//! remainder is collected as-is, without interpretation, into a generic
+//! `details` field.
+
+use super::string;
+use crate::native::block::DataType;
+use crate::native::packets::server::ColumnDescription;
+use crate::native::Error;
+use nom::bytes::streaming::is_not;
+use nom::bytes::streaming::tag;
+use nom::character::streaming::u64 as nom_u64;
+use nom::error::ErrorKind;
+use nom::sequence::delimited;
+use nom::sequence::separated_pair;
+use nom::sequence::terminated;
+use nom::Err as NomErr;
+use nom::IResult;
+
+/// Decode an array of `ColumnDescription`s from a buffer.
+pub fn decode(
+    src: &mut &[u8],
+) -> Result<Option<Vec<ColumnDescription>>, Error> {
+    // See `src/Storages/ColumnDescription.cpp` for details of the encoding
+    // here, but briefly:
+    //
+    // - The packet starts with a "header" describing the table name (usually
+    // empty), the version, the number of columns.
+    // - Each column is serialized as a string, with its name; type; comment;
+    // compression codec; settings; statistics; and TTL
+    // - These are all newline delimited, with tabs prefixing each of the items
+    // listed above.
+    //
+    // See https://github.com/ClickHouse/ClickHouse/blob/c82cf25b3e5864bcc153cbe45adb8c6527e1ec6e/src/Server/TCPHandler.cpp#L2433
+    // for more details, which is where the encoding of these values starts.
+    let Some(_table_name) = string::decode(src)? else {
+        return Ok(None);
+    };
+
+    // The column description itself is serialized as a ClickHouse
+    // varuint-prefixed string. Deserialize this, and then parse out the pieces
+    // of that parsed string.
+    let Some(text) = string::decode(src)? else {
+        return Ok(None);
+    };
+    column_descriptions(&text)
+}
+
+fn column_descriptions(
+    text: &str,
+) -> Result<Option<Vec<ColumnDescription>>, Error> {
+    // Try to match the version "header"
+    let text = match tag::<_, _, (_, ErrorKind)>("columns format version: 1\n")(
+        text,
+    ) {
+        Ok((text, _match)) => text,
+        Err(NomErr::Incomplete(_)) => return Ok(None),
+        Err(_) => return Err(Error::InvalidPacket("TableColumns (header)")),
+    };
+
+    // Match the number of columns.
+    let (text, n_columns) = match column_count(text) {
+        Ok(input) => input,
+        Err(NomErr::Incomplete(_)) => return Ok(None),
+        Err(_) => {
```

Comment:
I'm not really familiar with nom. Is it worthwhile to capture the actual error and report it as part of `InvalidPacket` ? I'm mainly thinking about how we debug parsing failures.

Reply:
Done in 997682132